File: diagonal-traverse.cpp
/*The code traverses the matrix diagonally, starting from the top-left corner and moving in an upward or downward direction alternatively.
It uses a boolean variable dir to determine the direction of movement. When dir is true, it moves upward, and when dir is false, it moves downward.
The result vector is filled with the elements in diagonal order, and the indices (r, c) are updated based on the current direction.
Time Complexity:
The code visits each element of the matrix exactly once, so the time complexity is O(m * n), where m and n are the dimensions of the matrix.
The space complexity is O(m * n) as the result vector stores all the elements in diagonal order.*/
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        int m=mat.size();
        int n=mat[0].size();
        vector<int> result(m * n);
        int r=0,c=0;
        bool dir=true;
        for(int i=0;i<m*n;i++){
            result[i]=mat[r][c];
            if(dir){
                if(r==0&&c!=n-1){
                    c++;
                    dir=false;
                }else if(c==n-1){
                    r++;
                    dir=false;
                }else{
                    r--;c++;
                }
            }else{
                if(c==0&&r!=m-1){
                    r++;
                    dir=true;
                }else if(r==m-1){
                    c++;
                    dir=true;
                }else{
                    r++;c--;
                }
            }
        }
        return result;
    }
};

File: diagonalTraverse.go
// Time Complexity : O(mn) m=rows and n=cols of matrix
// Space Complexity : O(1) (returned array space is not considered)
// Did this code successfully run on Leetcode : Yes

/*

Approach:
We play with the indices in order to get the resultant array. In this problem,
we have 2 directions we go in, either (bottom to top, left to right) or (top to
bottom, right to left). We keep a variable to track this direction, and keep 2
pointers to track the row and column indices. Then we introduce some checks to
consider boundary conditions.

*/

package main

func findDiagonalOrder(mat [][]int) []int {
	if len(mat) == 0 {
		return []int{}
	}

	m, n := len(mat), len(mat[0])
	dir := 1 // 1 for bottom to top, -1 for top to bottom
	res := make([]int, m*n)
	r, c := 0, 0

	for i := 0; i < m*n; i++ {
		res[i] = mat[r][c]

		// change R and C based on where we are in matrix
		// considering all constraints
		if dir == 1 { // while going bottom to top
			if c == n-1 {
				r++
				dir = -1
			} else if r == 0 {
				c++
				dir = -1
			} else {
				r--
				c++
			}
		} else if dir == -1 { // while going top to bottom
			if r == m-1 {
				c++
				dir = 1
			} else if c == 0 {
				r++
				dir = 1
			} else {
				r++
				c--
			}
		}
	}

	return res
}
